File: src/main/java/io/github/ajoz/workshop/fp/java/tools/seq/TakeFlow.java
package io.github.ajoz.workshop.fp.java.tools.seq;

import io.github.ajoz.workshop.fp.java.tools.Try;

import java.util.NoSuchElementException;

public final class TakeFlow<A> implements Flow<A> {
    private final Flow<A> upstream;
    private final int amount;

    private int taken;

    TakeFlow(final Flow<A> upstream,
             final int amount) {
        this.upstream = upstream;
        this.amount = amount;
    }

    @Override
    public Try<A> next() {
        // take an element from the upstream Flow
        final Try<A> next = upstream.next();
        // if the elemenent does not exist then just propagate the failure
        if (next.isFailure())
            return next;

        // if already taken enough elements then just propagate the failure
        if (taken >= amount) {
            return Try.failure(new NoSuchElementException("Reached the Flow amount: " + amount));
        }

        // increment the currently taken amount
        taken++;
        // return the upstream Flow item
        return next;
    }
}

File: kotlin/tools/src/main/kotlin/org/artrev/workshop/fp/tools/flow/TakeFlow.kt
package org.artrev.workshop.fp.tools.flow

import org.artrev.workshop.fp.tools.control.Try
import java.util.NoSuchElementException

fun <A> Flow<A>.take(amount: Int): Flow<A> {
    class TakeFlow(val upstream: Flow<A>,
                   val threshold: Int) : Flow<A> {
        private var taken: Int = 0

        override fun next(): Try<A> {
            // take an element from the upstream Flow
            val next = upstream.next()
            // if the element does not exist then just propagate the failure
            if (next.isFailure)
                return next

            // if already taken enough elements then just propagate the failure
            if (taken >= threshold) {
                return Try.Failure(NoSuchElementException("Reached the Flow take threshold: $threshold"))
            }

            // increment the currently taken amount
            taken++
            // return the upstream Flow item
            return next
        }
    }

    return TakeFlow(this, amount)
}
